Feature/merge 4.3

chtyim opened the pull request:
All conflict resolutions are done by picking the develop branch over the release branch.

Commit:
fix an error message

Commit:
[CDAP-13110] Fix dataset class loading in some places

albertshau committed:
CDAP-13129 ensure a sink is only ever in one mapreduce phase

Fixes a bug in the planner that could cause a sink to be present
in multiple phases of a pipeline. The planner will now insert a
connector in front of a sink if it has multiple reduce nodes
as parents, or if it has a reduce node and a source as a parent.

albertshau committed:
CDAP-13119 planner optimization to reduce connector datasets

Optimizing the planner to add a pass at the end that will merge
connectors if possible. Connectors can be merged if they have the
same branch head, which reduces the amount of I/O required while
still preserving the property that each record is only processed
once.

albertshau committed:
bump version to 4.3.4-SNAPSHOT

albertshau committed:
Merge pull request #9927 from caskdata/feature_release/CDAP-13119-planner-reduce-connectors

CDAP-13119 planner optimization to reduce connector datasets

albertshau committed:
fix checkstyle

albertshau committed:
Merge pull request #9928 from caskdata/fix-checkstyle

fix checkstyle

yaojiefeng committed:
CDAP-13123 fix stream exploration

albertshau committed:
CDAP-13122 use combine input format for connectors

This is a small optimization in case connector output is small
but in multiple files. In that scenario, this change will use
fewer mappers, lowering the overhead and resource usage of the
pipeline.

albertshau committed:
CDAP-13120 use correct directory for file in PartitionedFileSet

If there was no base path set for a PartitionedFileSet, it would
default to a 'files' directory whose parent is a directory named
after the dataset. This means when the dataset is deleted, only
the 'files' directory is deleted, but not the parent directory.
This behavior is especially problematic when using a
PartitionedFileSet as a local fileset in a workflow, as each run
of the workflow would leave a directory per PartitionedFileSet.

albertshau committed:
Merge pull request #9923 from caskdata/4-3-4-snapshot

bump version to 4.3.4-SNAPSHOT

albertshau committed:
Merge pull request #9930 from caskdata/bugfix_release/CDAP-13122-connector-input-format

CDAP-13122 use combine input format for connectors

yaojiefeng committed:
Merge pull request #9931 from caskdata/feature_release/CDAP-13123-fix-stream-explore

CDAP-13123 fix stream exploration

albertshau committed:
CDAP-13120 make defaulting base path backwards compatible

Modify the reconfigure logic so that the base path is not
defaulted unless it was defaulted when it was created. Also adding
unit tests.

ajainarayanan committed:
[CDAP-13111] Fixes performance of pipeline detailed view for a large pipeline + minor styling fix for run level section in pipeline detailed view

 - Fixes metrics for pipeline nodes to have proper ellipsis + renames a filter based on its function
 - Simplifies comma-separated-number filter to just add commas to a number (removed ellipsis adding functionality)

Commit:
Merge pull request #9921 from caskdata/bugfix_release/CDAP-13129-ensure-sink-in-single-mapred

CDAP-13129 ensure a sink is only ever in one mapreduce phase

albertshau committed:
CDAP-13120 fix unit tests

Commit:
fix build break

Commit:
Merge pull request #9937 from caskdata/feature/cdap-13110-fix-build

fix build break

ajainarayanan committed:
Merge pull request #9934 from caskdata/bugfix_release/ui-CDAP-13112

[CDAP-13111] - Fix performance issue with pipeline detailed view

albertshau committed:
Merge pull request #9932 from caskdata/bugfix_release/CDAP-13120-leftover-local-filesets

CDAP-13120 use correct directory for file in PartitionedFileSet

yaojiefeng committed:
log user who executes the program

ajainarayanan committed:
Fix references of node-uuid used for id for DOM elements

 - More patches to node uuids being set as ids for DOM elements
 - Fixes btn color for actions in dataset/stream detailed view in tracker

chtyim committed:
(CDAP-13158) Fix a race condition when running concurrent spark in workflow

Commit:
[CDAP-13171] Only set group on namespace directories if the group was explicitly configured

chtyim committed:
Merge pull request #9948 from caskdata/feature/CDAP-13158-fix-spark-conf-zip-race

(CDAP-13158) Fix a race condition when running concurrent spark in wo…

chtyim committed:
(CDAP-13179) Integrates TWILL-61 changes for AM resilience

ajainarayanan committed:
Merge pull request #9950 from caskdata/bugfix/ui-fix-node-uuids-for-dom-4.3.4

Fix references of node-uuid used for id for DOM elements

ajainarayanan committed:
[CDAP-13162] Retain code editors if macro'd in plugin configuration in pipelines

 - Fixes code editors in plugin config to not swap with input textboxes

ajainarayanan committed:
Merge pull request #9960 from caskdata/bugfix_release/ui-CDAP-13162

[CDAP-13162] Retain code editors if macro'd

sagarkapare committed:
CDAP-13205 Fix inconsistency when the Job is deleted but Trigger still exist in the schedulestore.

sagarkapare committed:
Merge pull request #9969 from caskdata/bugfix_release/cdap-13205

CDAP-13205 Fix inconsistency when the Job is deleted but Trigger still exist in the schedulestore.

Commit:
CDAP-13154 Use just the username and not kerberos principal for authorization

albertshau committed:
CDAP-13217 remove app spec check when recording run state

When we are recording run record state, we should not check for
existence of the application or program. Otherwise, there can
be a situation where an app is modified to remove a running
program. If that program is killed on the cluster, then the run
record will be in the 'running' state, which will eventually
be fixed by the run record corrector. However, when it is time
to save the run state, the program will not be found and the
state update will be ignored. This will loop infinitely.

albertshau committed:
CDAP-13191 fix bug in reading existing run records

Completed run records use a different rowkey than the other types
of run records. The previous code had a different and buggy method
to read existing run records, where the completed run record key
was missing the inverted timestamp part.

Fixed the code to use the working existing method to read a run
record for a given program run id.

albertshau committed:
rename method and mark argument as nullable

albertshau committed:
Merge pull request #9974 from caskdata/bugfix_release/CDAP-13217-remove-program-check-state-store

CDAP-13217 remove app spec check when recording run state

yaojiefeng committed:
Merge pull request #9945 from caskdata/feature_release/log-login-user-who-starts-the-program

log user who executes the program

chtyim committed:
Merge pull request #9956 from caskdata/feature/CDAP-13179-am-resilience

(CDAP-13179) Integrates TWILL-61 changes for AM resilience

albertshau committed:
Merge pull request #9975 from caskdata/bugfix_release/CDAP-13191-fix-completed-run-record-read

CDAP-13191 fix bug in reading existing run records

chtyim committed:
(CDAP-13218) Distinguish program state based on user actions

- Fix for Workflow, MR, and Spark that when user initiate a stop, emit KILLED instead of FAILED
- Ignore program run status event that won’t change the state in run record

Commit:
(CDAP-13223) Fixed single quotes in format strings

- Fixed format date directive to lose un-necessary single quote
- Removed on ENTER to submit a directive while concatenating string to a column
- Added a missing format, which caused the ordering between text and code to be messed up

chtyim committed:
Merge pull request #9977 from caskdata/feature/CDAP-13218-fix-workflow-state

(CDAP-13218) Distinguish program state based on user actions

bdmogal committed:
Merge pull request #9978 from caskdata/bugfix/format-date

(CDAP-13223) Fixed single quotes in format strings

Commit:
Merge pull request #9973 from caskdata/feature/CDAP-13154-auth-imp

CDAP-13154 Use just the username and not kerberos principal for auth…

chtyim committed:
Merge branch 'release/4.3' into feature/merge-4.3

chtyim commented:
Set of conflict files (minus **/pom.xml)

```
	both modified:   cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceAdmin.java
	both modified:   cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java
	both modified:   cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/DefaultStoreTest.java
	both modified:   cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/executor/DatasetAdminService.java
	both modified:   cdap-distributions/bin/build_docs_bucket.sh
	both modified:   cdap-distributions/src/emr/install.sh
	both modified:   cdap-distributions/src/hdinsight/pkg/install.sh
	both modified:   cdap-distributions/src/hdinsight/pkg/mainTemplate.json
	both modified:   cdap-distributions/src/packer/cdap-sdk-ubuntu16-with-uri.json
	both modified:   cdap-distributions/src/packer/files/cdap-sdk.json
	both modified:   cdap-examples/resources/weblog-analytics-config.json
	both modified:   cdap-examples/resources/weblog-analytics.txt
	both modified:   cdap-examples/resources/weblog-service.txt
	both modified:   cdap-explore/src/main/java/co/cask/cdap/explore/executor/ExploreExecutorHttpHandler.java
	both modified:   cdap-ui/app/cdap/components/FastAction/TruncateAction/index.js
	both modified:   cdap-ui/app/cdap/components/ProgramTable/index.js
	both modified:   cdap-ui/package.json
```

chtyim commented:
Build https://builds.cask.co/browse/CDAP-DUT6344

albertshau commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java:
```diff
@@ -773,7 +773,6 @@ private boolean isValid(RunRecordMeta existing, byte[] sourceId, String recordTy
       LOG.debug("Current source id '{}' is not larger than the existing source id '{}' in the existing " +
                   "run record meta '{}'. Skip recording program {}.",
                 Bytes.toHexString(sourceId), Bytes.toHexString(existingSourceId), existing, recordType);
-      return false;
```
this is still there on develop

chtyim replied:
oh yeah, missing that.

albertshau reviewed (commented):
one comment, otherwise lgtm

chtyim committed:
Fix the wrong conflict resolution and compilation error

albertshau committed:
fix race in creating app.meta dataset in test

Both the test and the service itself create the dataset. Use
getOrCreate so there is no conflict exception.

albertshau committed:
Merge pull request #9982 from caskdata/fix-race-in-test

fix race in creating app.meta dataset in test

chtyim committed:
Merge remote-tracking branch 'origin/release/4.3' into feature/merge-4.3

chtyim commented:
Build passed
